File: apps/go/internal/chapter10/randomforest.go
package chapter10

import (
	"fmt"
	"math/rand"
	"slices"

	"github.com/k2works/getting-started-machine-learning/apps/go/internal/chapter02"
	"github.com/k2works/getting-started-machine-learning/apps/go/internal/chapter03"
)

// unlimited は深さを制限しないことを表す値。
const unlimited = -1

// FittedTree は学習した決定木 1 本と、その木が使った特徴量の列・訓練データの行番号。
type FittedTree struct {
	Columns []string
	Rows    []int
	Model   *chapter03.DecisionTree
}

// RandomForest は第 3 章の決定木をブートストラップ標本と特徴量の部分集合で学習し、多数決で予測する。
type RandomForest struct {
	nEstimators int
	maxFeatures int
	maxDepth    int
	seed        int64
	trees       []FittedTree
}

// NewRandomForest は深さを制限しない決定木の森を返す。
func NewRandomForest(nEstimators, maxFeatures int, seed int64) *RandomForest {
	return &RandomForest{nEstimators: nEstimators, maxFeatures: maxFeatures, maxDepth: unlimited, seed: seed}
}

// NewRandomForestWithMaxDepth は深さの上限を指定した決定木の森を返す。
func NewRandomForestWithMaxDepth(nEstimators, maxFeatures, maxDepth int, seed int64) (*RandomForest, error) {
	if maxDepth < 0 {
		return nil, fmt.Errorf("深さの上限は 0 以上にしてください: %d", maxDepth)
	}

	return &RandomForest{nEstimators: nEstimators, maxFeatures: maxFeatures, maxDepth: maxDepth, seed: seed}, nil
}

// Trees は学習した決定木を返す。
func (r *RandomForest) Trees() []FittedTree {
	return r.trees
}

// MajorityVote はサンプルごとに、最も多い予測を選ぶ。同数なら先に現れた予測を選ぶ。
func MajorityVote(votes [][]string) []string {
	predictions := make([]string, len(votes[0]))

	for sample := range predictions {
		labels := make([]string, len(votes))
		for i, vote := range votes {
			labels[i] = vote[sample]
		}

		predictions[sample] = chapter03.Majority(labels)
	}

	return predictions
}

// BootstrapSample は 0 から size - 1 までの行番号を、重複を許して size 個選ぶ。
func BootstrapSample(size int, random *rand.Rand) []int {
	rows := make([]int, size)
	for i := range rows {
		rows[i] = random.Intn(size)
	}

	return rows
}

// SelectColumns は特徴量から、指定した列だけをその順に取り出す。
func SelectColumns(x []chapter02.Features, columns []string) ([]chapter02.Features, error) {
	selected := make([]chapter02.Features, len(x))

	for i, features := range x {
		values := make([]float64, len(columns))

		for j, column := range columns {
			value, err := features.Value(column)
			if err != nil {
				return nil, err
			}

			values[j] = value
		}

		row, err := chapter02.NewFeatures(columns, values)
		if err != nil {
			return nil, err
		}

		selected[i] = row
	}

	return selected, nil
}

// chooseColumns は列を並べ替えて先頭から maxFeatures 個選び、元の列の順に戻して返す。
func chooseColumns(allColumns []string, maxFeatures int, random *rand.Rand) []string {
	shuffled := slices.Clone(allColumns)
	random.Shuffle(len(shuffled), func(i, j int) { shuffled[i], shuffled[j] = shuffled[j], shuffled[i] })

	chosen := shuffled[:maxFeatures]
	columns := make([]string, 0, maxFeatures)

	for _, column := range allColumns {
		if slices.Contains(chosen, column) {
			columns = append(columns, column)
		}
	}

	return columns
}

// newTree は森の 1 本分の決定木を作る。
func (r *RandomForest) newTree() (*chapter03.DecisionTree, error) {
	if r.maxDepth == unlimited {
		return chapter03.Unlimited(), nil
	}

	return chapter03.WithMaxDepth(r.maxDepth)
}

// Fit はブートストラップ標本と特徴量の部分集合で、決定木を指定した本数だけ学習する。
func (r *RandomForest) Fit(x []chapter02.Features, t []string) error {
	if len(x) == 0 {
		return fmt.Errorf("特徴量が 1 件もありません")
	}

	if r.maxFeatures < 1 || r.maxFeatures > len(x[0].Columns) {
		return fmt.Errorf("使う特徴量の数が範囲外です: %d", r.maxFeatures)
	}

	random := rand.New(rand.NewSource(r.seed)) //nolint:gosec // 再現できる森のための擬似乱数で、暗号用途ではない
	allColumns := x[0].Columns
	trees := make([]FittedTree, 0, r.nEstimators)

	for range r.nEstimators {
		rows := BootstrapSample(len(x), random)
		columns := chooseColumns(allColumns, r.maxFeatures, random)

		sampleX := make([]chapter02.Features, len(rows))
		sampleT := make([]string, len(rows))

		for i, row := range rows {
			sampleX[i] = x[row]
			sampleT[i] = t[row]
		}

		selected, err := SelectColumns(sampleX, columns)
		if err != nil {
			return err
		}

		tree, err := r.newTree()
		if err != nil {
			return err
		}

		if err := tree.Fit(selected, sampleT); err != nil {
			return err
		}

		trees = append(trees, FittedTree{Columns: columns, Rows: rows, Model: tree})
	}

	r.trees = trees

	return nil
}

// Predict は木ごとの予測を多数決でまとめる。
func (r *RandomForest) Predict(x []chapter02.Features) ([]string, error) {
	if len(r.trees) == 0 {
		return nil, fmt.Errorf("Fit で学習してから Predict を呼んでください")
	}

	votes := make([][]string, len(r.trees))

	for i, tree := range r.trees {
		selected, err := SelectColumns(x, tree.Columns)
		if err != nil {
			return nil, err
		}

		vote, err := tree.Model.Predict(selected)
		if err != nil {
			return nil, err
		}

		votes[i] = vote
	}

	return MajorityVote(votes), nil
}

File: apps/ruby/lib/getting_started_ml/chapter10/random_forest.rb
# frozen_string_literal: true

module GettingStartedMl
  # 第 10 章のランダムフォレスト。第 3 章の決定木を束ねる。
  module Chapter10
    # ランダムフォレストの 1 本分。使った特徴量の列、ブートストラップ標本の行番号、学習した決定木。
    FittedTree = Data.define(:columns, :rows, :model)

    module_function

    # サンプルごとに、最も多い予測を選ぶ。同数なら先に現れた予測を選ぶ（第 3 章の majority と同じ）。
    def majority_vote(votes)
      return [] if votes.empty?

      votes.transpose.map { |labels| Chapter03.majority(labels) }
    end

    # 0 から size - 1 までの行番号を、重複を許して size 個選ぶ。
    def bootstrap_sample(size, rng)
      Array.new(size) { rng.rand(size) }
    end

    # 重複なしで count 個の列を選ぶ。選んだ列は元の順に並べ直す。
    def choose(columns, count, rng)
      columns & columns.sample(count, random: rng)
    end

    # 指定した列だけを取り出した特徴量にする。
    def select_columns(x, columns)
      x.map { |features| Chapter02::Features.new(columns:, values: columns.map { |column| features.value(column) }) }
    end

    # 自作のランダムフォレスト。木ごとにブートストラップ標本と特徴量の部分集合を選び、第 3 章の決定木を学習する。
    class RandomForest
      attr_reader :trees

      # max_depth が nil なら木の深さを制限しない。
      def initialize(n_estimators:, max_features:, max_depth: nil, seed: 0)
        @n_estimators = n_estimators
        @max_features = max_features
        @max_depth = max_depth
        @seed = seed
        @trees = nil
      end

      # 乱数の生成器を 1 つだけ作り、森全体で使い回す（木ごとに作り直すと全部の木が同じ標本になる）。
      def fit(x, t)
        rng = Random.new(@seed)
        @trees = Array.new(@n_estimators) { plant(x, t, rng) }
        self
      end

      # 全部の木の予測の多数決で予測する。
      def predict(x)
        raise "学習してから予測してください" if trees.nil?

        Chapter10.majority_vote(trees.map { |tree| tree.model.predict(Chapter10.select_columns(x, tree.columns)) })
      end

      private

      # 1 本分の標本と列を選び、決定木を学習する。
      def plant(x, t, rng)
        rows = Chapter10.bootstrap_sample(x.size, rng)
        columns = Chapter10.choose(x.first.columns, @max_features, rng)
        sample_x = Chapter10.select_columns(x.values_at(*rows), columns)
        model = Chapter03::DecisionTree.new(max_depth: @max_depth).fit(sample_x, t.values_at(*rows))

        FittedTree.new(columns:, rows:, model:)
      end
    end
  end
end
